[FIX] base: avoid infinite recursion on attachments search

agr-odoo opened the pull request:
Steps to reproduce:

Create a new local Odoo DB
Create another company on the DB
install the CRM module
Create an opportunity on the CRM app for each company.
Execute This server action (CHANGE THE res_id TO THE IDs OF THE OPPORTUNITY CREATED):
```
for _ in range(0, 300):
  for _ in range (0, 75):
    env['ir.attachment'].sudo().create({'name': 'test','res_model': 'crm.lead', 'res_id': 31, 'type': 'binary'})
  for _ in range (0, 75):
    env['ir.attachment'].sudo().create({'name': 'test','res_model': 'crm.lead', 'res_id': 45, 'type': 'binary'})
```
go to the Settings / Technical / Database Structure / Attachments.

Server will look in endless call because the search try to retrieve
`limit` records, but after applying company record rule we have
`limit - e`, so it begin a second search for e records.
Even the second search will look for limit and possibly will find less,
starting a third...

If the sub call is aware of how much records we actually need after
checking permission, we can return the recordset if we have collected
enough

opw-2369722

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/62293).

agr-odoo commented:
@nim-odoo 
Maybe a bit far-fetched, but the same code is used also in v14, so it may be worth fixing

agr-odoo commented:
Updated the naming to avoid confusion, `need` is the context variable used to know how many feasible record we need from the nested search
Updated also the recursive calculus of `need` (was forgotten before)

kitan191 reviewed (approved):
LGTM

@nim-odoo

nim-odoo reviewed (approved):
robodoo delegate+

agr-odoo commented:
@robodoo r+
